### ISSUE TITLE
Provide an AppImage for Linux distributions

### ISSUE THREAD

**Issue by Serkan-devel** [crmarsh/discord-linux-bugs#205]
(Based off of https://github.com/MrAlex94/Waterfox/issues/127)
Providing <a href="http://appimage.org/">Appimages</a> would have, among others, these advantages:
- Works for most Linux distributions (including Ubuntu, Fedora, openSUSE, CentOS, elementaryOS, Linux Mint, and others)
- One app = one file = super simple for users: just download one AppImage file, <a href="http://discourse.appimage.org/t/how-to-make-an-appimage-executable/80">make it executable</a>, and run
- No unpacking or installation necessary
- No root needed
- No system libraries changed
- Just one format for all major distributions
- Works out of the box, no installation of runtimes needed (unlike Snap and Flatpak)
- Optional desktop integration with `appimaged`
- Optional binary delta updates, e.g., for continuous builds (only download the binary diff) using AppImageUpdate
- Can optionally GPG2-sign your AppImages (inside the file)
- Works on Live ISOs (unlike Flatpak)
- Can use the same AppImages when dual-booting multiple distributions

<a href="https://github.com/probonopd/AppImageKit/wiki/AppImages">Here is an overview</a> of projects that are already distributing upstream-provided, official AppImages.

Would be nice if this is available here.

**Comment by crmarsh** [comment]
Sounds fine, I dunno why I wouldn't make a Snap and Flatpak at the same time, though.

**Comment by ahjolinna** [comment]
@crmarsh : thats true, I would personally wait to see which will become the more mainstream solution, plus those project aren't that mature yet  (more or less) they need some more development time.

this like the new init-system war ...and wayland vs MIR...so looking back to those, this will take a while for the linux community.
So until that day arrives discord should just provide at least basic rpm and deb, like chrome, viber, dropbox (and others) have done...oh and normal source pkgs would be good for other distros like arch.

also why not just use [OBS](http://openbuildservice.org/) to build all these pkgs, it supports many distros and it would also resolve the need for repo thing

**Comment by Serkan-devel** [comment]
Discord seems to be something that is written on electronjs and could probably be set as a build option https://github.com/lbryio/lbry-app/issues/48#issuecomment-295389466

But I almost completely agree with @ahjolinna 's comment because 
MIR is dead

**Comment by Svetomech** [comment]
I prefer Flatpak to AppImage because the latter provides no security whatsoever. Flatpak, on the other hand, runs in a [sandbox](https://github.com/projectatomic/bubblewrap).

**Comment by Serkan-devel** [comment]
While Appimages don't come with security by default (unless GPG-signing as stated in the first comment), it has naitive support for <a href="https://github.com/probonopd/AppImageKit/blob/master/README.md#sandboxing">low-overhead sandboxing</a> with <a href="https://github.com/netblue30/firejail/">firejail</a>.
[It is also possible to use the <a href="https://github.com/probonopd/AppImageKit/blob/master/sandbox/README.md">Bubblewrap</a> sandbox to run AppImages.]





/)

**Comment by Svetomech** [comment]
Can you show me at least one appimage that uses these _optional_ features? I never came across a single one.
Whereas with flatpaks security comes by default.

(\

**Comment by SapphireTactician** [comment]
I prefer just installing binaries as I do now. As long as this is an **alternative** and does not replace installation binaries, I'm fine with the idea.

**Comment by probonopd** [comment]
Yes, this is purely an _alternative_.

> Can you show me at least one appimage that uses these optional features?

Sandboxing is a _user_ option, i.e., the user can use Firejail which comes with native AppImage support. It is even possible to run AppImages using Bubblewrap, the same sandbox that Firejail uses.

**Comment by Serkan-devel** [comment]
*Flatpak

**Comment by Serkan-devel** [comment]
It's also like SELinux being optional in some cases.

**Comment by probonopd** [comment]
I have SELinux disabled all the times. Don't know how well AppImages work with it. Feedback welcome.

**Comment by simoniz0r** [comment]
Personally, I prefer AppImages to flatpaks or snaps because AppImages are extremely easy to work with on the user's end.  If built correctly, the user just has to `chmod a+x` the AppImage and run it.  The AppImage contains everything needed to run that wouldn't be present on a normal Linux system, so this would also be a good option for people who have troubles with the new `libc++` dependency.

**Comment by probonopd** [comment]
So what's the verdict here?

**Comment by Serkan-devel** [comment]
bump

**Comment by probonopd** [comment]
@simoniz0r how is updating implemented in these AppImages?

**Comment by simoniz0r** [comment]
Discord uses the user's config directory for downloading and applying updates, so that works fine.  It's just that the AppImage things Discord is done running when it closes between downloading and installing the updates that causes a slight annoyance.

**Comment by probonopd** [comment]
Do you know you will find a way around this? Or would it make more sense to use a standard update mechanism like [AppImageUpdate](https://github.com/AppImage/AppImageUpdate)?

**Comment by simoniz0r** [comment]
@probonopd No, it would not make sense to use AppImageUpdate.  Discord's updates usually happen through their servers; they will update some code on their end and push the update to the client without the client having to be reinstalled.

I'm honestly not sure how to find a way around this other than creating some janky bash script solution that would try to see if Discord is running and keep the AppImage running.

**Comment by probonopd** [comment]
The AppImage is supposed to be running (mounted) while the payload application is running. If it is not, then it is a bug and you will run into all sorts of issues anyway. So the first question is, why is the AppImage process exiting while something launched from within the AppImage is still running?

**Comment by simoniz0r** [comment]
@probonopd The AppImage is mounted when the application is running; Discord closes for a fraction of a second between downloading and installing updates, and this causes the AppImage to think that Discord is no longer running and delete the extracted directory.

**Comment by probonopd** [comment]
I see... we need to find a way around this.

**Comment by simoniz0r** [comment]
Yeah... like I said, I'm not really sure what I could do to prevent this on my end other than creating a bash script that would run a while loop, somehow check if Discord is running, sleep for like 15 seconds, and repeat the loop to keep the AppImage running when Discord closes to update.  That doesn't *really* sound like a great solution to me (if I could even figure out a reliable way to check if Discord is running) lol.

Depending on the version of Discord you're using, client updates aren't *too* frequent, so it's not really a show stopping bug right now; just annoying.

**Comment by probonopd** [comment]
>  AppImage's extracted dir in /tmp

Nothing is actually _extracted_ there, it's simply _mounted_ to a temporary mountpoint. What you could _try_ is have a custom `AppRun` that does something like this (untested!):

```
(...)
exec discord && sleep 2
```

This way, AppRun would be running 2 seconds longer than `discord`, hence keeping the mountpoint around for 2 more seconds. But it's kinda working around the original issue, which is:

Why is the discord process exiting prematurely while stuff launched by it is still executed. Maybe the original Discord authors can chime in here,

**Comment by simoniz0r** [comment]
>Nothing is actually extracted there, it's simply mounted to a temporary mountpoint.

Yeah, ok, you get my point lol.

>What you could try is have a custom AppRun that does something like this (untested!):

In order to solve this issue by sleeping after the Discord starts, I'd have to sleep for a very long time in order to make sure that it stays open long enough for the updates to be downloaded.

Honestly, like I said, I've had this issue with a few other applications that close between operations when trying to make AppImages out of them.  It'd be nice if this problem could be solved upstream instead of suggesting workarounds that would be very unreliable :-\

**Comment by probonopd** [comment]
How could the AppImage know how long it should stay mounted, if not by checking whether the AppRun process has exited?

**Comment by simoniz0r** [comment]
In my experience, any application that closes between operations closes for a fraction of a second.  If whatever is unmounting the AppImage were to wait even 2 seconds before doing that, it would solve every problem that I've had with creating AppImages for applications that close between operations.

**Comment by probonopd** [comment]
That is what `exec discord && sleep 2` would do...

**Comment by simoniz0r** [comment]
>That is what exec discord && sleep 2 would do...

Oh, yes, it would lol.  Not sure why, but I only saw one `&` the first time.  Still not really a fan of solutions like this, TBH.  It'd be nicer if the AppImage didn't instantly unmount itself every time that it thinks the application has finished running.

**Comment by simoniz0r** [comment]
@probonopd I just tested this a few times, and I'm not able to prevent the mounted AppImage directory from being unmounted by adding ` && sleep 2` to the exec line of my AppRun.  I even tried going up to 30 seconds, but I could see the mounted AppImage directory in `/tmp` being unmounted as soon as Discord closes between downloading updates and installing them.

**Comment by simoniz0r** [comment]
Just wanted to show that Discord doesn't throw any errors while downloading the updates, so it doesn't seem like Discord is crashing here; it just seems like the mounted AppImage directory is unmounted before it can relaunch.

![discord-appimage-console](https://u.teknik.io/manop.png)

**Comment by simoniz0r** [comment]
@probonopd Any more ideas?  It's going to seem to the average user like the Discord AppImages are just crashing every time an update from Discord's server is applied to the client.... don't really want to give them that impression.

**Comment by simoniz0r** [comment]
Fixed for now using this script inside of the AppImage to run Discord:

```bash
#!/bin/bash
./usr/share/discord-ptb/discord-ptb.wrapper && sleep 15
```

**Comment by Serkan-devel** [comment]
@simoniz0r do you consider `AppimageUpdate` to be also utilized on Discord?

**Comment by simoniz0r** [comment]
>do you consider AppimageUpdate to be also utilized on Discord?

Like I said above, no.  Most updates for Discord come from Discord's servers and don't require the client to be reinstalled, so it wouldn't work for those.  Updates that require the client to be reinstalled are much less frequent; PTB  and Stable have seen one update that required the client to be reinstalled in the entire time Discord has existed.

**Comment by Serkan-devel** [comment]
And what did the update feature to require a reinstall?

**Comment by simoniz0r** [comment]
>And what did the update feature to require a reinstall?

Not sure I understand what you're asking here?  Are you talking about PTB and Stable's move from 0.0.1 -> 0.0.2?  Things were changed in the client itself that couldn't be pushed as an update through there servers.  AFAIK, Discord wants to have the vast majority of updates happen in a way that the client does not need to be reinstalled.

**Comment by probonopd** [comment]
Scary. Self-Modding ever-changing code...
